File: desktop/src/shared/syncEngine.js
const EventEmitter = require('events');
const md5 = require('md5');
const { v4: uuidv4 } = require('uuid');
const fs = require('fs').promises;
const path = require('path');
const { app } = require('electron');

class SyncEngine extends EventEmitter {
  constructor(driveService, store) {
    super();
    this.driveService = driveService;
    this.store = store;
    this.syncStatus = 'idle';
    this.syncInterval = null;
    this.saveDebounceTimers = new Map();
    this.pandoricaFolderId = null;
    this.folderStructure = {};
  }

  async start() {
    try {
      // Initialize PANDORICA folder structure
      await this.initializeGoogleDriveStructure();

      // Initial sync
      await this.syncNow();

      // Start polling every 30 seconds
      this.syncInterval = setInterval(() => this.syncNow(), 30000);

      this.updateStatus('synced');
    } catch (error) {
      console.error('Sync start failed:', error);
      this.updateStatus('error');
    }
  }

  stop() {
    if (this.syncInterval) {
      clearInterval(this.syncInterval);
      this.syncInterval = null;
    }
  }

  updateStatus(status) {
    this.syncStatus = status;
    this.emit('status-change', status);
  }

  getSyncStatus() {
    return this.syncStatus;
  }

  async initializeGoogleDriveStructure() {
    // Create PANDORICA root folder
    const pandoricaFolder = await this.driveService.findOrCreateFolder('PANDORICA');
    this.pandoricaFolderId = pandoricaFolder.id;

    // Create subfolders
    const subfolders = ['daily', 'specs', 'bugs', 'ideas', 'archive'];
    for (const folderName of subfolders) {
      const folder = await this.driveService.findOrCreateFolder(folderName, this.pandoricaFolderId);
      this.folderStructure[folder.id] = {
        id: folder.id,
        name: folderName,
        type: 'folder'
      };
    }

    // Store in local store
    this.store.set('pandoricaFolderId', this.pandoricaFolderId);
    this.store.set('folderStructure', this.folderStructure);
  }

  async syncNow() {
    if (this.syncStatus === 'syncing') {
      return; // Already syncing
    }

    this.updateStatus('syncing');

    try {
      // Pull changes from Google Drive
      await this.pullFromDrive();

      // Push local changes to Google Drive
      await this.pushToDrive();

      this.updateStatus('synced');
    } catch (error) {
      console.error('Sync failed:', error);
      this.updateStatus('error');
    }
  }

  async pullFromDrive() {
    if (!this.pandoricaFolderId) {
      this.pandoricaFolderId = this.store.get('pandoricaFolderId');
      this.folderStructure = this.store.get('folderStructure', {});
    }

    // Get all files from Drive
    const driveFiles = await this.driveService.listFiles(this.pandoricaFolderId);
    const localNotes = this.store.get('notes', {});

    for (const driveFile of driveFiles) {
      if (driveFile.mimeType === 'application/vnd.google-apps.folder') {
        // Update folder structure
        if (!this.folderStructure[driveFile.id]) {
          this.folderStructure[driveFile.id] = {
            id: driveFile.id,
            name: driveFile.name,
            type: 'folder'
          };
        }
        continue;
      }

      // It's a file
      const localNote = localNotes[driveFile.id];

      if (!localNote) {
        // New file on Drive, download it
        const content = await this.driveService.getFile(driveFile.id);
        localNotes[driveFile.id] = {
          id: driveFile.id,
          name: driveFile.name,
          content: content,
          modifiedTime: new Date(driveFile.modifiedTime).getTime(),
          md5: driveFile.md5Checksum,
          synced: true
        };
      } else {
        // File exists locally, check for conflicts
        const conflict = await this.detectConflict(localNote, driveFile);

        if (conflict) {
          // Handle conflict
          await this.handleConflict(localNote, driveFile);
        } else if (driveFile.md5Checksum !== localNote.md5) {
          // Drive version is newer, update local
          const content = await this.driveService.getFile(driveFile.id);
          localNotes[driveFile.id] = {
            ...localNote,
            content: content,
            modifiedTime: new Date(driveFile.modifiedTime).getTime(),
            md5: driveFile.md5Checksum,
            synced: true
          };
        }
      }
    }

    // Check for deleted files
    for (const localId in localNotes) {
      const exists = driveFiles.find(f => f.id === localId);
      if (!exists && localNotes[localId].synced) {
        // File was deleted on Drive
        delete localNotes[localId];
      }
    }

    this.store.set('notes', localNotes);
    this.store.set('folderStructure', this.folderStructure);
  }

  async pushToDrive() {
    const localNotes = this.store.get('notes', {});
    const syncQueue = this.store.get('syncQueue', []);

    // Process sync queue (offline changes)
    for (const item of syncQueue) {
      try {
        if (item.action === 'create') {
          const fileData = await this.driveService.createFile(
            item.name,
            item.content,
            item.folderId,
            'text/markdown'
          );

          // Update local note with Drive ID
          localNotes[fileData.id] = {
            id: fileData.id,
            name: item.name,
            content: item.content,
            modifiedTime: new Date(fileData.modifiedTime).getTime(),
            md5: fileData.md5Checksum,
            synced: true
          };

          // Remove temporary local ID if exists
          if (item.tempId) {
            delete localNotes[item.tempId];
          }
        } else if (item.action === 'update') {
          const fileData = await this.driveService.updateFile(item.id, item.content);
          localNotes[item.id] = {
            ...localNotes[item.id],
            modifiedTime: new Date(fileData.modifiedTime).getTime(),
            md5: fileData.md5Checksum,
            synced: true
          };
        } else if (item.action === 'delete') {
          await this.driveService.deleteFile(item.id);
          delete localNotes[item.id];
        }
      } catch (error) {
        console.error('Sync queue item failed:', error);
        // Keep item in queue for retry
        continue;
      }
    }

    // Clear processed items from queue
    this.store.set('syncQueue', []);
    this.store.set('notes', localNotes);
  }

  async detectConflict(localNote, driveFile) {
    const localHash = md5(localNote.content);
    const remoteHash = driveFile.md5Checksum;

    if (localHash === remoteHash) {
      return null; // No conflict
    }

    const localTime = localNote.modifiedTime;
    const remoteTime = new Date(driveFile.modifiedTime).getTime();

    // If modified within 1 second and different content = conflict
    if (Math.abs(localTime - remoteTime) < 1000) {
      return {
        local: localNote,
        remote: driveFile,
        resolution: 'USER_CHOICE'
      };
    }

    return null;
  }

  async handleConflict(localNote, driveFile) {
    // For MVP: Last-write-wins with server time validation
    const serverTime = await this.driveService.getServerTime();
    const localTime = localNote.modifiedTime;
    const remoteTime = new Date(driveFile.modifiedTime).getTime();

    // Check for clock drift
    const drift = Math.abs(localTime - serverTime.getTime());
    if (drift > 300000) { // 5 minutes
      console.warn('Clock drift detected:', drift);
      // Use server time as source of truth
    }

    // Store conflict for user resolution
    const conflicts = this.store.get('conflicts', []);
    conflicts.push({
      id: uuidv4(),
      noteId: localNote.id,
      noteName: localNote.name,
      localContent: localNote.content,
      localTime: localTime,
      remoteContent: await this.driveService.getFile(driveFile.id),
      remoteTime: remoteTime,
      timestamp: Date.now()
    });

    this.store.set('conflicts', conflicts);
    this.emit('conflict-detected', conflicts[conflicts.length - 1]);

    // For now, keep both versions (user will resolve later)
    // Remote version wins by default
    const localNotes = this.store.get('notes', {});
    const remoteContent = await this.driveService.getFile(driveFile.id);
    localNotes[localNote.id] = {
      ...localNote,
      content: remoteContent,
      modifiedTime: remoteTime,
      md5: driveFile.md5Checksum,
      synced: true
    };
    this.store.set('notes', localNotes);
  }

  getFolders() {
    return Object.values(this.folderStructure);
  }

  getNotesInFolder(folderId) {
    const allNotes = this.store.get('notes', {});

    // For "All Notes" special folder
    if (folderId === 'all') {
      return Object.values(allNotes);
    }

    // Get notes from specific folder (would need parent tracking in real impl)
    // For MVP, return all notes
    return Object.values(allNotes);
  }

  async getNoteContent(noteId) {
    const notes = this.store.get('notes', {});
    return notes[noteId] ? notes[noteId].content : '';
  }

  async saveNote(noteId, content) {
    const notes = this.store.get('notes', {});

    if (!notes[noteId]) {
      return;
    }

    // Update local copy
    notes[noteId].content = content;
    notes[noteId].modifiedTime = Date.now();
    notes[noteId].md5 = md5(content);
    notes[noteId].synced = false;

    this.store.set('notes', notes);

    // Debounce sync
    if (this.saveDebounceTimers.has(noteId)) {
      clearTimeout(this.saveDebounceTimers.get(noteId));
    }

    const timer = setTimeout(() => {
      this.queueSync(noteId, 'update', content);
      this.saveDebounceTimers.delete(noteId);
    }, 500);

    this.saveDebounceTimers.set(noteId, timer);

    this.updateStatus('saving');
  }

  async createNote(folderId, title) {
    const tempId = 'temp-' + uuidv4();
    const notes = this.store.get('notes', {});

    const newNote = {
      id: tempId,
      name: title + '.md',
      content: `# ${title}\n\n`,
      modifiedTime: Date.now(),
      md5: md5(`# ${title}\n\n`),
      synced: false
    };

    notes[tempId] = newNote;
    this.store.set('notes', notes);

    // Queue for sync
    this.queueSync(tempId, 'create', newNote.content, folderId, title + '.md');

    return newNote;
  }

  async deleteNote(noteId) {
    const notes = this.store.get('notes', {});
    delete notes[noteId];
    this.store.set('notes', notes);

    // Queue delete
    this.queueSync(noteId, 'delete');
  }

  async createFolder(name) {
    const folder = await this.driveService.findOrCreateFolder(name, this.pandoricaFolderId);
    this.folderStructure[folder.id] = {
      id: folder.id,
      name: name,
      type: 'folder'
    };
    this.store.set('folderStructure', this.folderStructure);
    return folder;
  }

  async searchNotes(query) {
    const notes = this.store.get('notes', {});
    const results = [];

    for (const note of Object.values(notes)) {
      const content = note.content.toLowerCase();
      const name = note.name.toLowerCase();
      const searchTerm = query.toLowerCase();

      if (content.includes(searchTerm) || name.includes(searchTerm)) {
        // Find context
        const index = content.indexOf(searchTerm);
        const start = Math.max(0, index - 50);
        const end = Math.min(content.length, index + 50);
        const context = '...' + content.substring(start, end) + '...';

        results.push({
          id: note.id,
          name: note.name,
          context: context
        });
      }
    }

    return results;
  }

  queueSync(noteId, action, content = null, folderId = null, name = null) {
    const queue = this.store.get('syncQueue', []);
    queue.push({
      id: noteId,
      action: action,
      content: content,
      folderId: folderId,
      name: name,
      tempId: noteId.startsWith('temp-') ? noteId : null,
      timestamp: Date.now()
    });
    this.store.set('syncQueue', queue);

    // Trigger sync soon
    setTimeout(() => this.syncNow(), 1000);
  }
}

module.exports = SyncEngine;

File: mobile/PANDORICA/Services/SyncEngine.swift
import Foundation
import Combine
import CryptoKit

class SyncEngine: ObservableObject {
    @Published var syncStatus: SyncStatus = .idle
    @Published var notes: [Note] = []
    @Published var folders: [Folder] = []

    private var driveService: GoogleDriveService?
    private var syncTimer: Timer?
    private var pandoricaFolderId: String?
    private var folderStructure: [String: Folder] = [:]

    enum SyncStatus {
        case idle, syncing, synced, saving, error, offline
    }

    func start() {
        // Load from UserDefaults
        loadLocalData()

        // Start sync
        Task {
            await syncNow()
        }

        // Poll every 30 seconds
        syncTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            Task {
                await self?.syncNow()
            }
        }
    }

    func stop() {
        syncTimer?.invalidate()
        syncTimer = nil
    }

    func syncNow() async {
        await MainActor.run {
            syncStatus = .syncing
        }

        do {
            try await pullFromDrive()
            try await pushToDrive()

            await MainActor.run {
                syncStatus = .synced
            }
        } catch {
            print("Sync failed: \(error)")
            await MainActor.run {
                syncStatus = .error
            }
        }
    }

    func initializeGoogleDriveStructure(driveService: GoogleDriveService) async throws {
        self.driveService = driveService

        // Create PANDORICA root folder
        let pandoricaFolder = try await driveService.findOrCreateFolder(name: "PANDORICA")
        self.pandoricaFolderId = pandoricaFolder.id

        // Create subfolders
        let subfolders = ["daily", "specs", "bugs", "ideas", "archive"]
        for folderName in subfolders {
            let folder = try await driveService.findOrCreateFolder(name: folderName, parentId: pandoricaFolder.id)
            folderStructure[folder.id] = folder
        }

        await MainActor.run {
            self.folders = Array(folderStructure.values)
        }

        saveLocalData()
    }

    private func pullFromDrive() async throws {
        guard let driveService = driveService else { return }
        guard let pandoricaFolderId = pandoricaFolderId else { return }

        let driveFiles = try await driveService.listFiles(folderId: pandoricaFolderId)

        var updatedNotes = notes

        for driveFile in driveFiles {
            if driveFile.mimeType == "application/vnd.google-apps.folder" {
                // Update folder structure
                if folderStructure[driveFile.id] == nil {
                    folderStructure[driveFile.id] = Folder(
                        id: driveFile.id,
                        name: driveFile.name,
                        type: "folder"
                    )
                }
                continue
            }

            // It's a file
            if let existingIndex = updatedNotes.firstIndex(where: { $0.id == driveFile.id }) {
                let localNote = updatedNotes[existingIndex]

                // Check for conflicts
                if driveFile.md5Checksum != localNote.md5 {
                    let content = try await driveService.getFileContent(fileId: driveFile.id)
                    updatedNotes[existingIndex] = Note(
                        id: driveFile.id,
                        name: driveFile.name,
                        content: content,
                        modifiedTime: ISO8601DateFormatter().date(from: driveFile.modifiedTime ?? "") ?? Date(),
                        md5: driveFile.md5Checksum ?? "",
                        synced: true,
                        folderId: nil
                    )
                }
            } else {
                // New file on Drive
                let content = try await driveService.getFileContent(fileId: driveFile.id)
                updatedNotes.append(Note(
                    id: driveFile.id,
                    name: driveFile.name,
                    content: content,
                    modifiedTime: ISO8601DateFormatter().date(from: driveFile.modifiedTime ?? "") ?? Date(),
                    md5: driveFile.md5Checksum ?? "",
                    synced: true,
                    folderId: nil
                ))
            }
        }

        // Check for deleted files
        updatedNotes.removeAll { note in
            !driveFiles.contains { $0.id == note.id } && note.synced
        }

        await MainActor.run {
            self.notes = updatedNotes
            self.folders = Array(folderStructure.values)
        }

        saveLocalData()
    }

    private func pushToDrive() async throws {
        guard let driveService = driveService else { return }

        let unsyncedNotes = notes.filter { !$0.synced }

        for note in unsyncedNotes {
            if note.id.hasPrefix("temp-") {
                // Create new file
                let fileData = try await driveService.createFile(
                    name: note.name,
                    content: note.content,
                    parentId: note.folderId ?? pandoricaFolderId!
                )

                // Update local note
                if let index = notes.firstIndex(where: { $0.id == note.id }) {
                    await MainActor.run {
                        notes[index] = Note(
                            id: fileData.id,
                            name: note.name,
                            content: note.content,
                            modifiedTime: ISO8601DateFormatter().date(from: fileData.modifiedTime ?? "") ?? Date(),
                            md5: fileData.md5Checksum ?? "",
                            synced: true,
                            folderId: note.folderId
                        )
                    }
                }
            } else {
                // Update existing file
                let fileData = try await driveService.updateFile(fileId: note.id, content: note.content)

                if let index = notes.firstIndex(where: { $0.id == note.id }) {
                    await MainActor.run {
                        notes[index].md5 = fileData.md5Checksum ?? ""
                        notes[index].modifiedTime = ISO8601DateFormatter().date(from: fileData.modifiedTime ?? "") ?? Date()
                        notes[index].synced = true
                    }
                }
            }
        }

        saveLocalData()
    }

    // CRUD Operations
    func createNote(title: String, folderId: String?) {
        let tempId = "temp-" + UUID().uuidString
        let content = "# \(title)\n\n"
        let note = Note(
            id: tempId,
            name: "\(title).md",
            content: content,
            modifiedTime: Date(),
            md5: md5Hash(content),
            synced: false,
            folderId: folderId ?? pandoricaFolderId
        )

        notes.append(note)
        saveLocalData()

        Task {
            await syncNow()
        }
    }

    func saveNote(_ note: Note, content: String) {
        if let index = notes.firstIndex(where: { $0.id == note.id }) {
            notes[index].content = content
            notes[index].modifiedTime = Date()
            notes[index].md5 = md5Hash(content)
            notes[index].synced = false

            saveLocalData()

            Task {
                try? await Task.sleep(nanoseconds: 500_000_000) // 500ms debounce
                await syncNow()
            }
        }
    }

    func deleteNote(_ note: Note) {
        notes.removeAll { $0.id == note.id }
        saveLocalData()

        if let driveService = driveService, !note.id.hasPrefix("temp-") {
            Task {
                try? await driveService.deleteFile(fileId: note.id)
            }
        }
    }

    func searchNotes(query: String) -> [Note] {
        let lowercaseQuery = query.lowercased()
        return notes.filter { note in
            note.name.lowercased().contains(lowercaseQuery) ||
            note.content.lowercased().contains(lowercaseQuery)
        }
    }

    // Persistence
    private func saveLocalData() {
        if let notesData = try? JSONEncoder().encode(notes) {
            UserDefaults.standard.set(notesData, forKey: "notes")
        }

        if let foldersData = try? JSONEncoder().encode(folders) {
            UserDefaults.standard.set(foldersData, forKey: "folders")
        }

        if let pandoricaId = pandoricaFolderId {
            UserDefaults.standard.set(pandoricaId, forKey: "pandoricaFolderId")
        }
    }

    private func loadLocalData() {
        if let notesData = UserDefaults.standard.data(forKey: "notes"),
           let decodedNotes = try? JSONDecoder().decode([Note].self, from: notesData) {
            notes = decodedNotes
        }

        if let foldersData = UserDefaults.standard.data(forKey: "folders"),
           let decodedFolders = try? JSONDecoder().decode([Folder].self, from: foldersData) {
            folders = decodedFolders
            folderStructure = Dictionary(uniqueKeysWithValues: decodedFolders.map { ($0.id, $0) })
        }

        pandoricaFolderId = UserDefaults.standard.string(forKey: "pandoricaFolderId")
    }

    // Utility
    private func md5Hash(_ string: String) -> String {
        let data = Data(string.utf8)
        let hash = Insecure.MD5.hash(data: data)
        return hash.map { String(format: "%02hhx", $0) }.joined()
    }
}
